NBA teams are increasingly trotting out lineups with five players who can play and guard nearly any position. traditional positions don't accurately explain what a players skillset truly is, they incorrectly oversimplify the skill sets of NBA players. Simply plugging players into one of five positions does not accurately define a player’s specific skill set. Moreover, the misclassification of a player’s position may lead teams to waste resources on developing draft picks that do not fit their systems.

In light of these changes, we need an effective way to designate positions in the NBA not based on basic physical traits such as height and weight, but in terms of function, such as shooting and defense. A framework for modern NBA positions is important towards our understanding for how players have evolved, and effective roster construction. 

### Import the Required Python Packages and Methods

In [66]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import csv

### Finding and Evaluating Historical NBA Player Data

I will use the Kaggle dataset "NBA Players stats since 1950", with stats for all players since 1950. The file Seasons_Stats.csv contains the statics of all players since 1950, with details of each column as below:


    Year - Season
    Player - name
    Pos - Position
    Age - Age
    Tm - Team
    G - Games
    GS - Games Started
    MP - Minutes Played
    PER - Player Efficiency Rating
    TS% - True Shooting %
    3PAr - 3-Point Attempt Rate
    FTr - Free Throw Rate
    ORB% - Offensive Rebound Percentage
    DRB% - Defensive Rebound Percentage
    TRB% - Total Rebound Percentage
    AST% - Assist Percentage
    STL% - Steal Percentage
    BLK% - Block Percentage
    TOV% - Turnover Percentage
    USG% - Usage Percentage
    blanl
    OWS - Offensive Win Shares
    DWS - Defensive Win Shares
    WS - Win Shares
    WS/48 - Win Shares Per 48 Minutes
    blank2
    OBPM - Offensive Box Plus/Minus
    DBPM - Defensive Box Plus/Minus
    BPM - Box Plus/Minus
    VORP - Value Over Replacement
    FG - Field Goals
    FGA - Field Goal Attempts
    FG% - Field Goal Percentage
    3P - 3-Point Field Goals
    3PA - 3-Point Field Goal Attempts
    3P% - 3-Point Field Goal Percentage
    2P - 2-Point Field Goals
    2PA - 2-Point Field Goal Attempts
    2P% - 2-Point Field Goal Percentage
    eFG% - Effective Field Goal Percentage
    FT - Free Throws
    FTA - Free Throw Attempts
    FT% - Free Throw Percentage
    ORB - Offensive Rebounds
    DRB - Defensive Rebounds
    TRB - Total Rebounds
    AST - Assists
    STL - Steals
    BLK - Blocks
    TOV - Turnovers
    PF - Personal Fouls
    PTS - Points

In [67]:
with open('data/Seasons_Stats.csv','rb') as csvfile:
    reader1 = csv.reader(csvfile)
df1 = pd.read_csv('data/Seasons_Stats.csv')

First, we remove duplicated or empty rows and drop a couple of blank columns.

In [68]:
df1_no_duplicates = df1.drop_duplicates()
df2 = df1_no_duplicates
del df2['blanl']
del df2['blank2']
df2.dropna(how='all')
df2.head(5)

Unnamed: 0    Year           Player  Pos   Age   Tm     G  GS  MP  PER  \
0           0  1950.0  Curly Armstrong  G-F  31.0  FTW  63.0 NaN NaN  NaN   
1           1  1950.0     Cliff Barker   SG  29.0  INO  49.0 NaN NaN  NaN   
2           2  1950.0    Leo Barnhorst   SF  25.0  CHS  67.0 NaN NaN  NaN   
3           3  1950.0       Ed Bartels    F  24.0  TOT  15.0 NaN NaN  NaN   
4           4  1950.0       Ed Bartels    F  24.0  DNN  13.0 NaN NaN  NaN   

     TS%  3PAr    FTr  ORB%  DRB%  TRB%  AST%  STL%  BLK%  TOV%  USG%  OWS  \
0  0.368   NaN  0.467   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.1   
1  0.435   NaN  0.387   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  1.6   
2  0.394   NaN  0.259   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  0.9   
3  0.312   NaN  0.395   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.5   
4  0.308   NaN  0.378   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN -0.5   

   DWS   WS  WS/48  OBPM  DBPM  BPM  VORP     FG    FGA    FG%  3P  3PA  3P%  \
0  3.6  3.5    NaN   NaN   NaN  NaN   NaN  144.0  516.0  0.279 NaN  NaN  NaN   
1  0.6  2.2    NaN   NaN   NaN  NaN   NaN  102.0  274.0  0.372 NaN  NaN  NaN   
2  2.8  3.6    NaN   NaN   NaN  NaN   NaN  174.0  499.0  0.349 NaN  NaN  NaN   
3 -0.1 -0.6    NaN   NaN   NaN  NaN   NaN   22.0   86.0  0.256 NaN  NaN  NaN   
4 -0.1 -0.6    NaN   NaN   NaN  NaN   NaN   21.0   82.0  0.256 NaN  NaN  NaN   

      2P    2PA    2P%   eFG%     FT    FTA    FT%  ORB  DRB  TRB    AST  STL  \
0  144.0  516.0  0.279  0.279  170.0  241.0  0.705  NaN  NaN  NaN  176.0  NaN   
1  102.0  274.0  0.372  0.372   75.0  106.0  0.708  NaN  NaN  NaN  109.0  NaN   
2  174.0  499.0  0.349  0.349   90.0  129.0  0.698  NaN  NaN  NaN  140.0  NaN   
3   22.0   86.0  0.256  0.256   19.0   34.0  0.559  NaN  NaN  NaN   20.0  NaN   
4   21.0   82.0  0.256  0.256   17.0   31.0  0.548  NaN  NaN  NaN   20.0  NaN   

   BLK  TOV     PF    PTS  
0  NaN  NaN  217.0  458.0  
1  NaN  NaN   99.0  279.0  
2  NaN  NaN  192.0  438.0  
3  NaN  NaN   29.0   63.0  
4  NaN  NaN   27.0   59.0

A second file, players.csv, contains static information for each player, as height, weight, etc.

In [69]:
players = pd.read_csv('data/Players.csv', index_col=0)
players.head(5)

Player  height  weight                          collage    born  \
0  Curly Armstrong   180.0    77.0               Indiana University  1918.0   
1     Cliff Barker   188.0    83.0           University of Kentucky  1921.0   
2    Leo Barnhorst   193.0    86.0         University of Notre Dame  1924.0   
3       Ed Bartels   196.0    88.0  North Carolina State University  1925.0   
4      Ralph Beard   178.0    79.0           University of Kentucky  1927.0   

    birth_city birth_state  
0          NaN         NaN  
1     Yorktown     Indiana  
2          NaN         NaN  
3          NaN         NaN  
4  Hardinsburg    Kentucky

I then selected these three data categories:
    
    Player - Player's full name (first and last)
    height - Height in cm
    weight - Weight in kg

1.The players have unique names (checked at the beginning), we can merge the two dataframes using the 'Player' column.

In [70]:
data = pd.merge(df2, players[['Player', 'height', 'weight']], left_on='Player', right_on='Player', right_index=False,
      how='left', sort=False)
#.fillna(value=0)

2.Using rate statistics (i.e. points per game) or cumulative statistics (i.e. total points) can be misleading when it comes to analysis because these statistics tend to inflate players with lengthier careers. To deal with outliers, I instituted a minimum threshold of 40 games played. Also I keep only players with more than 400 minutes for each season (with a 82 games regular season, thats around 5 minutes per game. Players with less than that will be only anecdotical, and will distort the analysis).

In [71]:
data = data[~(data['Pos']==0) & (data['MP'] > 400) & (data['G'] > 40)]
data.reset_index(inplace=True, drop=True)

3.Replace the * sign in some of the names.

In [72]:
data['Player'] = data['Player'].str.replace('*','')

4.For the stats that represent total values (others, as TS%, represent percentages), we will take the values per 36 minutes. The reason is to judge every player according to his characteristics, not the time he was on the floor.

In [73]:
totals = ['PER', 'OWS', 'DWS', 'WS', 'OBPM', 'DBPM', 'BPM', 'VORP', 'FG', 'FGA', '3P', '3PA', '2P', '2PA', 'FT', 'FTA',
         'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']

for col in totals:
    data[col] = 36 * data[col] / data['MP']

In [75]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15383 entries, 0 to 15382
Data columns (total 53 columns):
Unnamed: 0    15383 non-null int64
Year          15383 non-null float64
Player        15383 non-null object
Pos           15383 non-null object
Age           15381 non-null float64
Tm            15383 non-null object
G             15383 non-null float64
GS            11192 non-null float64
MP            15383 non-null float64
PER           15365 non-null float64
TS%           15383 non-null float64
3PAr          11668 non-null float64
FTr           15383 non-null float64
ORB%          12997 non-null float64
DRB%          12997 non-null float64
TRB%          13543 non-null float64
AST%          14260 non-null float64
STL%          12997 non-null float64
BLK%          12997 non-null float64
TOV%          12165 non-null float64
USG%          12165 non-null float64
OWS           15365 non-null float64
DWS           15365 non-null float64
WS            15365 non-null float64
WS/48   

from the data information above, the ideal number of data rows we want to work with is 15383, most amount of data is missing from 

    3P  (3-Point Field Goals)                  11668 non-null float64
    3PA (3-Point Field Goal Attempts)          11668 non-null float64
    3P% (3-Point Field Goal Percentage)        10597 non-null float64
    3PAr(3-Point Attempt Rate)                 11668 non-null float64
    
It's becasue 3-point line was not introduced to NBA until 1979, but for the sake of classifying modern players, I have to assume all players before 1979 have not attempted 3-pointers, so I will fill the missing values of these three columns with 0. 

In addition:

    GS  (Games Started)                        11192 non-null float64

Since number of Games started for each player does not effect their performance measuring, I will delete this column.

In [76]:
threePointsCol = ['3P','3PA','3P%','3PAr']
for feature in threePointsCol:
    data[feature] = data[feature].fillna(0)
del data['GS']
data[threePointsCol].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15383 entries, 0 to 15382
Data columns (total 4 columns):
3P      15383 non-null float64
3PA     15383 non-null float64
3P%     15383 non-null float64
3PAr    15383 non-null float64
dtypes: float64(4)
memory usage: 480.8 KB


The next groups of columns that have substantical amount of missing data is from:
    
    ORB           12997 non-null float64
    DRB           12997 non-null float64
    STL           12997 non-null float64
    BLK           12997 non-null float64
    TOV           12165 non-null float64
    
And their associate percentages:
    
    ORB%          12997 non-null float64
    DRB%          12997 non-null float64
    STL%          12997 non-null float64
    BLK%          12997 non-null float64
    TOV%          12165 non-null float64
    
Since most of the above data has minimul 12997 rows, the commonality is due to the fact ORB, DRB, STL, BLK and their associate percentages are not included in this dataset before 1974, these missing values cannot be filled with assumption, but at the same time these columns are too important to be removed from the construction of player classification, therefore all the data before 1974 have to be discarded. 

In [77]:
data = data[~(data['Year'] < 1974)]

5.Cast these columns to be of type 'int'.

In [82]:
integerCol = ['Year','Age', 'G']
for feature in integerCol:
    data[feature] = data[feature].astype(dtype ='int')

In [88]:
pd.options.display.max_columns = None
data.sample(5).transpose()

14226              6586          14118       3359   \
Unnamed: 0         22735               9978         22543        5212   
Year                2014               1990          2014        1978   
Player      John Salmons  Dominique Wilkins  Jarrett Jack     Ron Lee   
Pos                   SG                 SF            PG          PG   
Age                   34                 30            30          25   
Tm                   TOR                ATL           CLE         PHO   
G                     60                 80            80          82   
MP                  1281               2888          2252        1928   
PER             0.213583           0.306648      0.183837     0.30249   
TS%                0.472              0.556         0.498       0.478   
3PAr               0.409              0.109         0.265           0   
FTr                0.152               0.34         0.213        0.24   
ORB%                 1.1                8.4           1.1           5   
DRB%                 9.6               12.4          10.3         8.6   
TRB%                 5.3               10.4           5.6         6.8   
AST%                12.1               11.8          22.7        21.8   
STL%                 1.4                2.2           1.3           5   
BLK%                 0.7                  1           0.8         0.5   
TOV%                11.2                8.3          14.8        17.4   
USG%                12.6               30.4          17.8        24.5   
OWS           0.00562061           0.115928     0.0143872           0   
DWS            0.0337237          0.0224377     0.0223801   0.0709544   
WS             0.0393443           0.138366     0.0367673   0.0709544   
WS/48              0.052              0.184         0.049       0.094   
OBPM          -0.0618267          0.0660665    -0.0223801 -0.00186722   
DBPM           0.0028103         -0.0199446    -0.0143872   0.0392116   
BPM           -0.0562061          0.0461219    -0.0367673   0.0373444   
VORP                   0          0.0523546   -0.00319716   0.0373444   
FG               3.06323             10.097       4.57194     7.78631   
FGA               8.3185            20.8421       11.1581     17.7386   
FG%                0.368              0.484          0.41       0.439   
3P               1.32084           0.735457        1.0071           0   
3PA              3.40047            2.28116       2.95737           0   
3P%                0.388              0.322         0.341           0   
2P               1.74239             9.3615       3.56483     7.78631   
2PA              4.91803            18.5609       8.20071     17.7386   
2P%                0.354              0.504         0.435       0.439   
eFG%               0.448              0.502         0.455       0.439   
FT                0.9274            5.72161       1.99822     3.17427   
FTA              1.26464             7.0928       2.38188     4.25726   
FT%                0.733              0.807         0.839       0.746   
ORB             0.337237            2.70499      0.351687     1.77386   
DRB              2.95082            3.78947       3.21314     2.96888   
TRB              3.28806            6.49446       3.56483     4.74274   
AST              2.89461            2.49307        5.1794     5.69502   
STL             0.983607            1.57064      0.895204     4.20124   
BLK             0.309133           0.585873      0.351687    0.317427   
TOV              1.12412            2.16898       2.12611     4.12656   
PF               2.58548            1.75762       2.12611     4.79876   
PTS              8.37471             26.651       12.1492     18.7469   
height               201                201           190         193   
weight                95                 90            90          87   

                  14534  
Unnamed: 0        23258  
Year               2015  
Player      Mike Miller  
Pos                  SF  
Age                

In [80]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12997 entries, 2386 to 15382
Data columns (total 52 columns):
Unnamed: 0    12997 non-null int64
Year          12997 non-null float64
Player        12997 non-null object
Pos           12997 non-null object
Age           12997 non-null float64
Tm            12997 non-null object
G             12997 non-null float64
MP            12997 non-null float64
PER           12997 non-null float64
TS%           12997 non-null float64
3PAr          12997 non-null float64
FTr           12997 non-null float64
ORB%          12997 non-null float64
DRB%          12997 non-null float64
TRB%          12997 non-null float64
AST%          12997 non-null float64
STL%          12997 non-null float64
BLK%          12997 non-null float64
TOV%          12165 non-null float64
USG%          12165 non-null float64
OWS           12997 non-null float64
DWS           12997 non-null float64
WS            12997 non-null float64
WS/48         12997 non-null float64
OBPM 